File: backend/internal/schedule/lesson.go
package schedule

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

type DayOfWeek int

const (
	Monday    DayOfWeek = 1
	Tuesday   DayOfWeek = 2
	Wednesday DayOfWeek = 3
	Thursday  DayOfWeek = 4
	Friday    DayOfWeek = 5
	Saturday  DayOfWeek = 6
	Sunday    DayOfWeek = 7
)

func (d DayOfWeek) Valid() bool { return d >= Monday && d <= Sunday }

func FromTime(t time.Time) DayOfWeek {
	wd := t.Weekday()
	if wd == time.Sunday {
		return Sunday
	}
	return DayOfWeek(int(wd))
}

// TimeOfDay represents a clock time (HH:MM) stored as minutes since midnight.
// Using a compact integer avoids timezone/DST confusion for recurring weekly slots.
type TimeOfDay int

func ParseTimeOfDay(s string) (TimeOfDay, error) {
	t, err := time.Parse("15:04", s)
	if err != nil {
		t, err = time.Parse("15:04:05", s)
		if err != nil {
			return 0, fmt.Errorf("schedule: invalid time %q: %w", s, err)
		}
	}
	return TimeOfDay(t.Hour()*60 + t.Minute()), nil
}

func (t TimeOfDay) String() string { return fmt.Sprintf("%02d:%02d", int(t)/60, int(t)%60) }
func (t TimeOfDay) Valid() bool    { return t >= 0 && t < 24*60 }

type Lesson struct {
	ID          uuid.UUID
	ClassroomID uuid.UUID
	Subject     string
	TeacherID   *uuid.UUID
	DayOfWeek   DayOfWeek
	StartsAt    TimeOfDay
	EndsAt      TimeOfDay
	Notes       string
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

func (l Lesson) Overlaps(other Lesson) bool {
	if l.DayOfWeek != other.DayOfWeek {
		return false
	}
	return l.StartsAt < other.EndsAt && other.StartsAt < l.EndsAt
}

File: mobile/lib/shared/models/lesson.dart
class Lesson {
  final String id;
  final String classroomId;
  final String subject;
  final String? teacherId;
  final int dayOfWeek;
  final String startsAt;
  final String endsAt;
  final String? notes;
  final String createdAt;
  final String updatedAt;

  const Lesson({
    required this.id,
    required this.classroomId,
    required this.subject,
    this.teacherId,
    required this.dayOfWeek,
    required this.startsAt,
    required this.endsAt,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Lesson.fromJson(Map<String, dynamic> json) => Lesson(
        id: json['id'] as String,
        classroomId: json['classroomId'] as String,
        subject: json['subject'] as String,
        teacherId: json['teacherId'] as String?,
        dayOfWeek: json['dayOfWeek'] as int,
        startsAt: json['startsAt'] as String,
        endsAt: json['endsAt'] as String,
        notes: json['notes'] as String?,
        createdAt: json['createdAt'] as String,
        updatedAt: json['updatedAt'] as String,
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'classroomId': classroomId,
        'subject': subject,
        if (teacherId != null) 'teacherId': teacherId,
        'dayOfWeek': dayOfWeek,
        'startsAt': startsAt,
        'endsAt': endsAt,
        if (notes != null) 'notes': notes,
        'createdAt': createdAt,
        'updatedAt': updatedAt,
      };
}
